### ISSUE TITLE
Implement new two-tier subprocess API

### ISSUE THREAD

**Issue by lucacasonato** [denoland/deno#11016]
```ts
declare namespace Deno {
  export interface ExecOptions {
    cmd: string[] | [URL, ...string[]];
    cwd?: string;
    env?: Record<string, string>;
    stdin?: "piped" | "inherit" | "null" | Deno.File; // default is "null"
    stdout?: "piped" | "inherit" | "null" | Deno.File; // default is "piped"
    stderr?: "piped" | "inherit" | "null" | Deno.File; // default is "piped"
  }

  export interface ExecOutput {
    status: ProcessStatus;
    stdout?: Uint8Array; // if it was set to "piped" in options, this is set
    stderr?: Uint8Array; // if it was set to "piped" in options, this is set
  }

  /**
   * Start a subprocess and wait until it completes. The subprocess will be
   * signaled to exit automatically if this process dies.
   */
  export function exec(options: ExecOptions): Promise<ExecOutput>;

  // NOTE: `Child` has multiple underlying resources:
  //          1. stdio - these have to always be closed manually
  //          2. process itself - this is closed automatically when `status`
  //             resolves (so after a `.kill` or after natural death).
  export class Child {
    readonly pid: number;

    readonly stdin?: Writer & Closer; // if it was set to "piped" in options, this is set
    readonly stdout?: Reader & Closer; // if it was set to "piped" in options, this is set
    readonly stderr?: Reader & Closer; // if it was set to "piped" in options, this is set

    readonly status: Promise<ProcessStatus>;

    kill(code: number): void;
  }

  export interface SpawnOptions extends ExecOptions {
    detached?: boolean;
  }

  /**
   * Start a subprocess in the background. The returned `Child` handle can be
   * used to communicate with the parent.
   */
  export function spawn(options: SpawnOptions): Child;
}
```

This new API provides a high level, super easy to use API (`Deno.exec`), and a low-level more intricate API (`Deno.spawn`). `Deno.exec` can be used for all cases where the user just wants to spawn a subprocess and wait for it to complete. For this API no resources need to be manually closed. `Deno.spawn` can be used for longer lived subprocess that you communicate with - they can also be detached, so outlive their parent.

The new semantics for backing resources are as follows:

- No resources need to be manually closed for `Deno.exec`
- For `Deno.spawn` "piped" stdio needs to be closed manually. The process itself is also a resource, but it is automatically cleaned up after the process exits.

The semantics for process lifecycle:

- By default we try to make sure a process does not outlive its parent. This is done by putting processes into process groups, jobs, or using setsid.
- A user can opt to set a process to be "detached" in which case we will ensure the subprocess can outlive its parent.

**Comment by wperron** [comment]
In its current form, the current proposal would be a No for me.

I like that we're adding `Deno.file` to the stdin, stdout and stderr options,
though I would prefer having `Reader & Closer` for stdin and `Writer & Closer`
for stdout and stderr as that would be more flexible and also represent the
direction of the stream better. Since `Deno.File` implements all of `Reader`,
`Writer` and `Closer`, file handles would still be able to be used just like the
current proposal.

Where this proposal loses me is in the distinction between `exec` and `spawn`.
It seems like the distinction is fairly arbitrary; under the hood, both spawn
a child process after all. Given that the underlying operation is the same, it
feels weird to have different data structures to represent them -- why couldn't
the output of `exec` return the same data structure as the output of `spawn`?
The `status` property is a `ProcessStatus` on `ExecOutput` but is a
`Promise<ProcessStatus>` on the `Child` class; That's needlessly confusing in my
opinion. Further, `exec` returns a `Promise` of a simple object with a defined
shape whereas `spawn` returns an instance of a class that has state. That
difference is there to accomodate the fact that we have two different proposed
functions that behave in very different ways.

Instead, I think we can improve the proposal by instead having `Deno.run` be a
constructor for a `Command` type rather than automatically start the underlying
process. We can attach keep the input parameter of `Deno.run` similar to the
proposed `ExecOptions` interface, supplying the same arguments to it (maybe also
include the changes detailed in the first paragraph) and attach high-level
control functions like `start`, `run`, `wait` onto that class. It might make
sense then to rename `Deno.run` to `Deno.cmd`. A simple example of a command to
just fire-and-forget would look like this:

```typescript
let cmd = await Deno.cmd({
  cmd: ["cat", "./this-file.txt"],
  stdout: Deno.stdout,
}).run();

console.log(`exited with code ${cmd.status});
```

We can also easily add lower level APIs by embedding a `Process` struct as a
public property of `Command` that represents the lower-level knowledge of what's
happening at the OS level. This is where we'd stick functions like `detach`,
`pid`, `exited` and others. An example of a program that starts a zombie process
would look like this:

```typescript
let cmd = Deno.cmd({
  cmd: ["deno", "run", "--allow-net", "--allow-read", "https://deno.land/std@0.99.0/http/file_server.ts", dir],
  stdout: Deno.stdout,
});
cmd.process.detach();
cmd.start();
Deno.exit(0);
```

As I see it, this approach offers a smaller API surface area for the easy case
while also exposing the switches and knobs for the lower-level, more complex
case. There's fewer top-level options and types to remember, and the `Deno.cmd`
function behaves the same whether the intent is to have a long-running process
or not. Tentatively, the interfaces would look a bit like this:

```typescript
declare namespace Deno {
  export interface CommandOptions {
    cmd: string[] | [URL, ...string[]];
    cwd?: string;
    env?: Record<string, string>,
    stdin?: "piped" | "inherit" | "null" | Deno.Reader & Deno.Closer;  // default "null"
    stdout?: "piped" | "inherit" | "null" | Deno.Writer & Deno.Closer; // default "null"
    stderr?: "piped" | "inherit" | "null" | Deno.Writer & Deno.Closer; // default "null"
  }

  export class Command {
    cmd: string[] | [URL, ...string[]];
    cwd?: string;
    env?: Record<string, string>,
    stdin?: "piped" | "inherit" | "null" | Deno.Reader & Deno.Closer;  // default "null"
    stdout?: "piped" | "inherit" | "null" | Deno.Writer & Deno.Closer; // default "null"
    stderr?: "piped" | "inherit" | "null" | Deno.Writer & Deno.Closer; // default "null"
    process?: Process // undefined if the process has not yet started or already exited

    start(): void;     // starts a process in the background
    wait(): void;      // blocks until the program has exited (maybe should be async?)
    async run(): void; // starts a process and resolves when it has exited
  }

  export class Process {
    pid?: number;

    detach(): void;
    wait(): void;
    kill(): void;
    signal(): void;
  }
}
```

**Comment by nayeemrmn** [comment]
> ```ts
>     readonly stdin?: Writer & Closer; // if it was set to "piped" in options, this is set
>     readonly stdout?: Reader & Closer; // if it was set to "piped" in options, this is set
>     readonly stderr?: Reader & Closer; // if it was set to "piped" in options, this is set
> ```

These properties being optional in `Deno.Process` first lead to users always having to use `!` assertions, then a lot of complexity with generic types to alleviate that problem. I think they should be present unconditionally and throw a runtime `TypeError` when you try to read/write them. Not sure what to do with `ExecOutput`.

**Comment by wperron** [comment]
> > ```ts
> >     readonly stdin?: Writer & Closer; // if it was set to "piped" in options, this is set
> >     readonly stdout?: Reader & Closer; // if it was set to "piped" in options, this is set
> >     readonly stderr?: Reader & Closer; // if it was set to "piped" in options, this is set
> > ```
> 
> These properties being optional in `Deno.Process` first lead to users always having to use `!` assertions, then a lot of complexity with generic types to alleviate that problem. I think they should be present unconditionally and throw a runtime `TypeError` when you try to read/write them. ~Not sure what to do with `ExecOutput`~ `ExecOutput::{stdout,stderr}` should similarly have throwing getters.

could also remove the `?` since there's only 4 possible types with default values

**Comment by eibens** [comment]
If you care for feedback on this from a user: 

I recently built a [wrapper for Deno.run](https://deno.land/x/jog) that does something similar to the proposed `Deno.exec` function. It takes a `Uint8Array | string` for `stdin`, and returns `stdout` as a `Uint8Array`. If the process exits with an error, the contents of `stderr` are thrown. A mapping function can be used to transform `stdiout` into a string, or parse it. I realize that this is making a lot of assumptions, but it works for plenty situations. 

The main benefits of such an interface, at least to me, are I can get process output with a single instruction and that it prevents me from having to figure out for each new project which resources need to be closed and in what order.

**Comment by jsejcksn** [comment]
> > ```ts
> >     readonly stdin?: Writer & Closer; // if it was set to "piped" in options, this is set
> >     readonly stdout?: Reader & Closer; // if it was set to "piped" in options, this is set
> >     readonly stderr?: Reader & Closer; // if it was set to "piped" in options, this is set
> > ```
> 
> These properties being optional in `Deno.Process` first lead to users always having to use `!` assertions, then a lot of complexity with generic types to alleviate that problem. I think they should be present unconditionally and throw a runtime `TypeError` when you try to read/write them. ~Not sure what to do with `ExecOutput`~ `ExecOutput::{stdout,stderr}` should similarly have throwing getters.

I am against this suggested pattern of throwing on property access because it cannot be communicated in the type system (`never` gets absorbed by any other type in a union), which leaves the full burden of communicating this on JSDoc comments or other documentation, which is generally less accessible.

`undefined` is the correct type for a property without a value.
